modules/programs/mercurial.nix

```diff
@@ -0,0 +1,104 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.mercurial;
+
+in
+
+{
+
+  options = {
+    programs.mercurial = {
+      enable = mkEnableOption "Mercurial";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.mercurial;
+        defaultText = "pkgs.mercurial";
+        description = "Mercurial package to install.";
+      };
+
+      userName = mkOption {
+        type = types.str;
+        description = "Default user name to use.";
+      };
+
+      userEmail = mkOption {
+        type = types.str;
+        description = "Default user email to use.";
+      };
+
+      aliases = mkOption {
+        type = types.attrs;
+        default = {};
+        description = "Mercurial aliases to define.";
+      };
+
+      extraConfig = mkOption {
+        type = types.either types.attrs types.lines;
+        default = {};
+        description = "Additional configuration to add.";
+      };
+
+      iniContent = mkOption {
+        type = types.attrsOf types.attrs;
+        internal = true;
+      };
+
+      ignores = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        example = [ "*~" "*.swp" ];
+        description = "List of globs for files to be globally ignored.";
+      };
+
+      regexpIgnores = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        example = [ "^.*~$" "^.*\\.swp$" ];
+        description =
+            "List of regular expressions for files to be globally ignored.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable (
+    mkMerge [
+      {
+        home.packages = [ cfg.package ];
+
+        programs.mercurial.iniContent.ui = {
+          username = cfg.userName + " <" + cfg.userEmail + ">";
+        };
+
+        xdg.configFile."hg/hgrc".text = generators.toINI {} cfg.iniContent;
+
+      }
+
+      (mkIf (cfg.ignores != [] || cfg.regexpIgnores != []) {
+
+        programs.mercurial.iniContent.ui.ignore =
+            "${config.xdg.configHome}/hg/hgignore_global";
+
+        xdg.configFile."hg/hgignore_global".text = 
+            "syntax: glob\n"   + concatStringsSep "\n" cfg.ignores + "\n" +
+            "syntax: regexp\n" + concatStringsSep "\n" cfg.regexpIgnores + "\n";
```

Comment:
I don't know anything about configuring Mercurial, just want to verify that it is OK with the generated configuration if either `ignores` or `regexpIgnores` is empty.

Reply:
The `"syntax: *"` lines just operate as switches, so this should be absolutely fine as far as the hgignore file is concerned. It obviously it lacks elegance though and leads to a slightly messy file if `ignores` is empty.

Would happily adopt a neater suggestion - I decided to keep it this way for now because this way if the user writes directely to `xdg.configFile."hg/hgignore_global".text` elsewhere, they will get the default behaviour (which is to use regular expression syntax).

The `"syntax: regexp"` flag should appear regardless, so maybe something like the below (untested) - although that leads to a somewhat uglier module for questionable benefit:
```
xdg.configFile."hg/hgignore_global".text = 
   (if cfg.ignores != []
    then  "syntax: glob\n"   + concatStringsSep "\n" cfg.ignores + "\n"
    else "") +
    "syntax: regexp\n" + concatStringsSep "\n" cfg.regexpIgnores + "\n";
```